### PR TITLE
inventory: Handle IndexError while parsing limit file

### LINKED ISSUE #59695: ERROR! Unexpected Exception, this is probably a bug: string index out of range
##### SUMMARY
Ansible run fails and I want to retry the failed hosts with the --limit @windows.retry command and it fails with "ERROR! Unexpected Exception, this is probably a bug: string index out of range"

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

limit

##### ANSIBLE VERSION

```paste below
$ ansible --version
ansible 2.9.0.dev0
  config file = /Users/tanner/projects/ansible.git/playbooks.git/celadonsystems.com/ansible.cfg
  configured module search path = [u'/Users/tanner/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/tanner/projects/ansible.git/ansible/lib/ansible
  executable location = /Users/tanner/projects/ansible.git/ansible/bin/ansible
  python version = 2.7.10 (default, Feb 22 2019, 21:55:15) [GCC 4.2.1 Compatible Apple LLVM 10.0.1 (clang-1001.0.37.14)]
```

##### CONFIGURATION

```paste below
$ ansible-config dump --only-changed

```

##### OS / ENVIRONMENT

Management host macOS 10.14.6
Managed hosts various releases of Windows

##### STEPS TO REPRODUCE

Run ansible and have a failure so you get the retry file
```yaml
$ ansible-playbook -i testing ../windows.yml  --limit @../windows.retry
```

##### EXPECTED RESULTS
Ansible will run for just the failed hosts in the retry file

##### ACTUAL RESULTS
ERROR! Unexpected Exception, this is probably a bug: string index out of range

```paste below
ERROR! Unexpected Exception, this is probably a bug: string index out of range
```

### PR DESCRIPTION
**Pull request by Akasurde**
##### SUMMARY
Handle IndexError exception raised while parsing the limit file.

Fixes: #59695

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME
changelogs/fragments/limit_file_parsing.yml
lib/ansible/inventory/manager.py
test/integration/targets/inventory/runme.sh

### CODE REVIEW

File: lib/ansible/inventory/manager.py [hunk 1]
```diff
@@ -69,9 +69,9 @@ def order_patterns(patterns):
     pattern_intersection = []
     pattern_exclude = []
     for p in patterns:
-        if p[0] == "!":
```
Comment (bcoca) [inline lib/ansible/inventory/manager.py]: instead of adding p on every condition, `if not p: continue` would shortcut the whole thing

**Review by bcoca — APPROVED** [review]

File: lib/ansible/inventory/manager.py [hunk 2]
```diff
@@ -69,9 +69,9 @@ def order_patterns(patterns):
...
+        if p and p[0] == "!":
```
Comment (mkrizek) [inline lib/ansible/inventory/manager.py]: @sivel  So this was introduced in https://github.com/ansible/ansible/pull/58400 and there are more changes where we replaced `startswith` with checking the first index instead for perf purposes. I wonder if doing `p and p[0]` over `startswith` is still beneficial. Or, if we should filter empty patterns someplace up, in the code that calls this.

Comment (bcoca) [reply lib/ansible/inventory/manager.py]: @mkrizek see my comment above, checking `if not p` would still be a net gain

Comment (mkrizek) [reply lib/ansible/inventory/manager.py]: @bcoca +1, I should've refreshed before commenting ;-)

Still, not for this PR but we should audit https://github.com/ansible/ansible/pull/58400 for similar possibilities of `IndexError`.

**Review by bcoca — APPROVED** [review]

**Pull request merged**